ambry-store/src/main/java/com/github/ambry/store/BlobStore.java

```diff
@@ -835,6 +845,15 @@ private void maybeCallBeforeSynchronizationFunc() {
     }
   }
 
+  /**
+   * Call {@link #betweenGetEndOffsetAndFindKeyFunc} if it's not null. This is for testing only.
+   */
+  private void maybeCallBetweenGetEndOffsetAndFindKeyFunc() {
+    if (betweenGetEndOffsetAndFindKeyFunc != null) {
+      betweenGetEndOffsetAndFindKeyFunc.run();
```

Comment:
For this type of conditional execution, AtomicObject is safer than volatile.  II believe it's technically possible for this to get a NPE since volatile doesn't prevent the variable from being overwritten between checking for null and running.

Reply:
good catch on the NPE. But I think it's fine to use volatile since we are not doing any CAS operation here. I updated to make sure the NPE doesn't happen.